### ISSUE TITLE
Parallelize: Send `func`, `args` and `kwargs` once to each process using pool initializer

### ISSUE THREAD

**Issue by kelvinburke** [sciris/sciris#621]
When creating a Pool, each process can run a custom `initializer` function where we can pass the function `func` and the `args` and `kwargs` which don't change from task to task. These can be set to global values for that process so that the subsequent tasks can run the function without it having to be passed for each task.

This can be useful when pickling, passing and unpickling the function and args takes longer than (or similar time to) the task itself.

There is one thing to be careful of:
If we pass the pickled args to each process, unpickle them and set them to a global value, we can reuse them for each task. Then, if the function modifies the values of the args, this could lead to unpredictable results as currently each task gets its own deepcopied version of the args.

So:
By default the method can be:
 - pickle the function and args
 - send those to the new processes through the pool creation `initializer`
 - they save the pickled function and args to a global value (global for that process)
 - for each task the function and args get unpickled and gets called with the task-specific args added
 
This will slightly speed up all map calls, assuming that pickling and sending to the new process through the `initializer` (which I think pickles again) is the same speed as the built-in pickling that occurs when doing `pool.map`. The speed up comes from now having to not pickle and send the function and args each time.
 
But, if we know that each task doesn't need a new deepcopy of the function and args, then we can reuse them. Giving the following:
 - pickle the function and args
 - send those to the new processes through the pool creation `initializer`
 - they unpickle the function and args and save the unpickled to a global value (global for that process)
 - for each task call the function with task-specific args added
 This could be `parallelizer='XX-nocopy'` and would be much faster than currently.

 I can share some example code that I've been playing with if you like.

**Comment by kelvinburke** [comment]
@cliffckerr One thing to note is that `multiprocess and multiprocessing` don't necessarily give different copies of an `arg` to each task:
```
import sciris as sc
def func(a, *args, **kwargs):
    print('id(a)', id(a), list(a.keys()))
    a['C'] = 'WOW'
    return
if __name__ == "__main__":
    num = 20
    ncpus = 3
    a = dict()
    sc.parallelize(func, kwargs=dict(a=a), parallelizer='multiprocess', iterkwargs=dict(b=range(num)), ncpus=ncpus)
```
You get:
```
id(a) 2343018634304 []
id(a) 2343018634304 ['C']
id(a) 1492483646848 []
id(a) 1936019955712 []
id(a) 1492483646848 ['C']
id(a) 1936019955712 ['C']
id(a) 2342957809536 []
id(a) 2342957809536 ['C']
id(a) 1492480293184 []
id(a) 1492480293184 ['C']
id(a) 1936017330752 []
id(a) 1936017330752 ['C']
id(a) 2343018642560 []
id(a) 2343018642560 ['C']
id(a) 1492483731136 []
id(a) 1492483731136 ['C']
id(a) 1936015503552 []
id(a) 1936015503552 ['C']
id(a) 2343018933120 []
id(a) 2343018933120 ['C']
```
This is because `map` splits the arguments into chunks given by `chunksize`, and shared objects in different tasks will get shared if they are in the same chunk. This can be disabled by doing `map_func = partial(map_func, chunksize=1)`. So `multiprocess and multiprocessing` are kind of doing the same thing as I am suggesting, but for each chunk, not just once for each Process.

So, either it is not important that each task gets its own deepcopy of the arguments/function and we do the 2nd method (where the process unpickles it straight away and reuses it for every task it gets), or, by default we do the 1st method which ensures each task gets its own deepcopy and the 2nd method is an option if you know it won't break anything.

I am equally content with this being switched off by default but then turning on this "pre-sending" of the `func, args, kwargs` with a parallelizer option:
 - `"multiprocess"`  - which shares the args within the chunk (current, pickles and unpickles once per chunk)
 - `"multiprocess-presend-copy"`  - which will pre-send the args and unpickle a new copy for each task (faster in that it only pickles them once, but slower because unpickles more than current)
 - `"multiprocess-presend-share"` - which will pre-send the args and unpickle once, sharing the args for each task in that process (fastest because pickle once, unpickle once per process)

**Comment by RomeshA** [comment]
An alternative that is sometimes feasible is to have a global/shared variable in the workspace before dispatching the parallel job, then when the process is forked (on Unix) each worker will have its own copy of that variable for free. Similar sort of thing can be achieved on Windows by having the script create the variable when imported, although it's a bit slower because Windows runs the script on every worker at startup. The limitation of course is that this only occurs when the workers are created, so it's not as flexible if you're reusing the pool of workers for different tasks

**Comment by kelvinburke** [comment]
Yes, but when on Unix and the process is forked, all the process share that global variable (there is only one copy). 
I can remember times when **I think** this has caused a slow down?
If they get that shared value for free then we could on starting the new process, just make our own copy of the shared value possibly? This is similar to what I have suggested, but no need to pickle to begin with, just copy the free copy.

**Comment by kelvinburke** [comment]
See my implemented using `parallelizer='X-presend-copy' or 'X-presend-share'` https://github.com/kelvinburke/sciris/tree/presend-parallel-func

**Comment by kelvinburke** [comment]
Noting that it adds another dependency `cloudpickle` which appears to be the same or slightly more robust than `dill` at pickling lambdas etc, and quicker than `dill`. I haven't checked how efficient it is in terms of size of pickled object.

**Comment by cliffckerr** [comment]
Finally got around to looking at this, sorry for the delay @kelvinburke! I think it's a good idea but would need to think about it a bit more ... it's so easy to make unintentional regressions w.r.t. parallelization that I'd want to make sure it's tested carefully on different platforms (especially Windows vs Linux, which has been a source of trouble for us in the past). I do like the idea of `cloudpickle` though, we run into problems with lambda functions all the time, and if this solves that, that would be great! Would want to check performance too though...
